python/fusion_engine_client/analysis/file_index.py

```diff
@@ -0,0 +1,139 @@
+from typing import Union
+
+import os
+
+import numpy as np
+
+from ..messages import MessageType, Timestamp
+
+
+class FileIndex(object):
```

Comment:
This file could use usage comments.

If I understand the high level usage, this is basically a wrapper around a np array of `DTYPE` that serializes to `RAW_DTYPE` for output to a file. Everything else are basically helper functions for getting/building this array.

Reply:
I'm a little confused why `RAW_DTYPE` is ever expected to be used outside of this class. I would try to simplify this constructor by not having construction from `RAW_DTYPE` be an option. 

Basically, just simplifying this interface to support the expected usage pattern, which appears to be either:
- Load from file
- Iteratively built from FileIndexBuilder
- Made from a filtered subset of another FileIndex.

Of these it seems like the index_path version of this constructor is really the only one that would be used outside of this file. In that case the logic for construction from a list could live in  FileIndexBuilder for example.

Reply:
Yeah that makes sense. I removed the `RAW_DTYPE` construction -- that's really an internal detail to save/load -- and made the dtype variables protected. This class still uses `RAW_DTYPE` internally for saving and loading so we can't move it entirely to the builder class. Otherwise this class would have to defer to the builder on every file load. The builder's purpose is really just for building up the data to create an index when one doesn't already exist, but it doesn't deal with any of the raw serialization stuff.